### PR TITLE
add mixerRouteConfig cache

### PR DESCRIPTION
**Pull request by irisdingbj**
1. add a cache to prevent protobuf <-> json converting

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -78,6 +79,9 @@ type PushContext struct {
 	ServicePort2Name map[string]PortList `json:"-"`
 
 	initDone bool
+
+	// MixerRouterFilterConfig has mixer Router Filter config, indexed by service hostname + outbound/inbound
+	MixerRouterFilterConfig map[string]*types.Struct `json:"-,omitempty"`
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go]: I  do not find where this field is updated and cleared.

**Comment by hzxuzhonghu** [comment]
This optimization can works well, @rshriram should we put it here. 

Since we have 
```
	// All data is set when the PushContext object is populated in `InitContext`,
	// data should not be changed by plugins.
```

**Comment by irisdingbj** [comment]
@kyessenov This is just is used to improve performance found in https://github.com/istio/istio/issues/9800 so no function change. Based on this I think no new test is need instead we need evaluate performance improvement with this change.

**Comment by kyessenov** [comment]
@irisdingbj I understand that there should not by any functional change. What evidence do we have it's the case? Looking at https://codecov.io/gh/istio/istio/src/master/pilot/pkg/networking/plugin/mixer/mixer.go, I see no coverage for the modified code. One thing you can do is to achieve high unit test coverage first, store the golden artifacts, and then validate that the code preserves the output.

**Comment by irisdingbj** [comment]
@kyessenov Thanks for the codecov pointer. I found out the original test coverage for this mixer.go is not good so most of the code has no testing coverage from the beginning.

File: pilot/pkg/model/push_context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (irisdingbj) [reply pilot/pkg/model/push_context.go]: Thanks @mandarjog  Done~

**Comment by irisdingbj** [comment]
@mandarjog Can you please /approve this change if no other issues? Thanks~

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 2]
```diff
@@ -344,16 +347,48 @@ func buildInboundTCPFilter(mesh *meshconfig.MeshConfig, node *model.Proxy, attrs
...
+		config = &mccpb.ServiceConfig{
+			DisableCheckCalls: disableClientPolicyChecks(in.Env.Mesh, in.Node),
+			MixerAttributes:   &mpb.Attributes{Attributes: attrs},
+			ForwardAttributes: &mpb.Attributes{Attributes: attrs},
+		}
+	} else {
+		config = buildInboundRouteConfig(push, in, in.ServiceInstance)
+	}
 	// These settings are not backward compatible with 0.8.
 	// Proxy version is only available from 1.0 onwards.
```
Comment (costinm) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: Can we remove this ? 1.1 doesn't need compat with 0.8 ( in separate PR maybe)

Comment (irisdingbj) [reply pilot/pkg/networking/plugin/mixer/mixer.go]: OK, I will remove this for 1.1 in a new PR. Thanks!

**Comment by mandarjog** [comment]
@costinm re `I'm not very happy with push context changes outside of 'before the push'` 
 Please advise on where is the appropriate place to do it ?
From the looks of it, the correct approach is to precompute the cache for all destinations and dispense with all these locks.

We are in no hurry to get this in, so if there is a better way, we should use the better way.
I don't want to add to the tech debt.

**Comment by irisdingbj** [comment]
@mandarjog Have a talk with @costinm "if you can init them before the push even better. But I think 'on-demand' may be ok Since long term we'll have the envoy on-demand So not everything can be pre-computed"  So it should be Ok for the change.

File: pilot/pkg/model/push_context.go [hunk 3]
```diff
@@ -81,6 +82,9 @@ type PushContext struct {
 	ServicePort2Name map[string]PortList `json:"-"`
 
 	initDone bool
+
+	// MixerPerRouteFilterConfig has mixer Router Filter config, indexed by service hostname + outbound/inbound
+	MixerPerRouteFilterConfig map[string]*types.Struct `json:"-"`
```
Comment (rshriram) [inline pilot/pkg/model/push_context.go]: My comment didn't make it through. But can you please send a follow up PR to remove this and move it into the plugins? There is no point in spilling plugin state back into the core leading to all the same issues we had in 0.7.

Comment (rshriram) [reply pilot/pkg/model/push_context.go]: Add a precompute callback to all the plugins.. during initialization, call configgen.BuildSharedState which in turn should call all the precompute callbacks and cache the state in the config gen.

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 4]
```diff
@@ -344,16 +347,48 @@ func buildInboundTCPFilter(mesh *meshconfig.MeshConfig, node *model.Proxy, attrs
...
-	if _, found := node.GetProxyVersion(); !found {
+	if _, found := in.Node.GetProxyVersion(); !found {
 		config.ForwardAttributes = nil
 	}
-	filterConfigs[mixer] = util.MessageToStruct(config)
+	filterConfig = util.MessageToStruct(config)
+	filterConfigs[mixer] = filterConfig
+	push.Mutex.Lock()
+	if push.MixerPerRouteFilterConfig == nil {
+		push.MixerPerRouteFilterConfig = map[string]*types.Struct{}
```
Comment (rshriram) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: This is a very very bad idea to update the cache right from the plugin. You end up serializing everything (every proxy will block on this lock/unlock during the first connect. See how the CDS caching is done and please follow the same.

**Pull request merged**